### DISCUSSION TITLE
How to validate provided fixed values using @validate_call?

### DISCUSSION THREAD

**Discussion by cringelord000222** [pydantic/pydantic#8974] [Question]
Hi there, 

Below is a sample function, I want to allow only value `int:3` or `int:17` for the argument `user_input`, something like `user_input: Annotated[int, Validator(values=[3,17])]` (just an example) but I'm not sure how to do it exactly, especially with the `validate_call` decorator, most examples that I could get from the internet are implementations of BaseModel and using literals. 

How should I edit my function signature so that the `user_input` argument will check if all the input values are `int` and either `3` or `17`? If the input is anything other than this 2 value, it will return ValidateError, and something like `"Input should be 3 or 17 as specified ..."`.

I need a way to enforce my own set of values, regardless of int or str, if posssible. Thanks.

```
@validate_call
def do_something(user_input: Annotated[int, "this argument only allows either <3> or <17> as input"] = 3) -> str:
```

**Comment by Viicos** [comment]
For literal values, [`Literal`](https://docs.python.org/3/library/typing.html#typing.Literal) will work: `user_input: Literal[3, 17]`. Are you looking to do the same for non literal values?

**Reply by cringelord000222** [reply]
Yup, I was able to get the error `Input should be 3 or 17 [type=literal_error, input_value=1, input_type=int]` with `stack_index: Literal[3, 17],`. But I would like to know the way for non-literal values too, as sometimes we need to explicitly declare the value type.

**Reply by Viicos** [reply]
In that case, that would mean you want to validate against a specific instance of a class, are you looking for something like this?:

```python
class MyClass:
    pass

my_inst = MyClass()

@validate_call
def func(arg: Literal[my_inst]):  # not valid, but I'm assuming this is what you are looking for
    # this assertion should be valid:
    assert arg is my_inst
```

**Reply by cringelord000222** [reply]
Hi there, I gave your code a try but I can't get it to validate the attributes for my instance. 

In the example below I had `"adam"` as an input when I have declared `Literal["brian", "sally"]` in my init function, but I guess I did it in the wrong part, and the code just runs like usual and didn't throw any errors. My assumption was the func() would be able to validate my instance attributes. Can you make corrections for me? 

Thanks

```
from pydantic import validate_call
from typing import Literal


class MyClass:
    def __init__(self, name: Literal["brian", "sally"], num: int) -> None:
        self.name = name
        self.num = num


my_inst = MyClass(name="adam", num=1)


@validate_call
def func(
    arg: Literal[my_inst],
):  # not valid, but I'm assuming this is what you are looking for
    # this assertion should be valid:
    assert arg is my_inst


func(arg=my_inst)

print(my_inst.__dict__)
`````

**Reply by Viicos** [reply]
> the attributes for my instance.

What do you mean by that? Could you provide a minimal example of what you are trying to achieve?

**Reply by cringelord000222** [reply]
The top example code that I've given is just an example of me trying to validate my instance (I had `name` and `num` in my instance), but unfortunately I can't make it working (if you run the snippet i've given it doesn't give error). And I need some help on where do I type hint my instance's attributes, so that the above code gives error. 

I have declared `name: Literal["brian", "sally"]` in my init function, so I expect an error during instance validation, because I created my object using `name="adam"`.

**Reply by Viicos** [reply]
I missed a part of your example. Now I think I understand better. Your `MyClass` is a vanilla Python class, you won't have any validation from Python here. A type checker on the other hand will be able to catch these errors. If you want runtime validation, well this is what Pydantic is for:

```python
class MyClass(BaseModel):
    name: Literal["brian", "sally"]
    num: int

MyClass(name="invalid", num=1)
#> ValidationError thrown
```

**Reply by cringelord000222** [reply]
Hi, yes, this example is what I found from the internet too, but everyone is using "BaseModel" way to validate instances, is there a sample implementation that uses the `"validation_call"` decorator instead?
